src/pymor/algorithms/lradi.py

```diff
@@ -193,31 +199,34 @@ def projection_shifts(A, E, V, prev_shifts):
         The |Operator| E from the corresponding Lyapunov equation.
     V
         A |VectorArray| representing the currently computed iterate.
+    Z
+        A |VectorArray| representing the current approximate solution.
     prev_shifts
         A |NumPy array| containing the set of all previously used shift
         parameters.
+    shift_options
+        The shift options to use (see :func:`lyap_lrcf_solver_options`).
 
     Returns
     -------
     shifts
         A |NumPy array| containing a set of stable shift parameters.
     """
-    if prev_shifts[-1].imag != 0:
-        Q = gram_schmidt(cat_arrays([V.real, V.imag]), atol=0, rtol=0)
+    if shift_options['subspace_columns'] == 1:
+        if prev_shifts[-1].imag != 0:
+            Q = gram_schmidt(cat_arrays([V.real, V.imag]), atol=0, rtol=0)
+        else:
+            Q = gram_schmidt(V, atol=0, rtol=0)
     else:
-        Q = gram_schmidt(V, atol=0, rtol=0)
+        num_columns = shift_options['subspace_columns'] * len(V)
+        Q = gram_schmidt(Z[-num_columns:], atol=0, rtol=0)
 
-    Ap = A.apply2(Q, Q)
-    Ep = E.apply2(Q, Q)
-
-    shifts = spla.eigvals(Ap, Ep)
-    shifts.imag[abs(shifts.imag) < np.finfo(float).eps] = 0
-    shifts = shifts[np.real(shifts) < 0]
+    shifts = spla.eigvals(A.apply2(Q, Q), E.apply2(Q, Q))
```

Comment:
have you considered the implicit computation of what was formerly called `Ap`? It can be reconstructed from `Z` and the previous shifts. 

Still, this explicit handling sometimes gives better results anyway, so it is probably fine to be a bit more expensive here.

Reply:
No, I just did the simplest things here to improve the FOM example. Adding an option for implicit projection might be good, but I would not implement it in this PR.

Reply:
agreed